Small oracle changes

bonomat opened the pull request:
A few minor changes I made while trying to understand the bug in #2002 .

bonomat committed:
HTTP requests should time out eventually

10 seconds is arbitrarily chosen but should be plenty for fetching from our oracle.

luckysori commented on daemon/src/oracle.rs:
```diff
@@ -199,6 +199,7 @@ impl Actor {
 
                     let response = client
                         .get(url.clone())
+                        .timeout(core::time::Duration::from_secs(10))
```
❓ Could it hang forever before?

luckysori replied:
🔧 Maybe extract the `Duration` into a constant so that we can explain the reason, even if it's fairly arbitrary.

bonomat replied:
> ❓ Could it hang forever before?

My understanding is yes and I couldn't find any evidence that it wouldn't in the reqwest codebase.

luckysori commented on daemon/src/oracle.rs:
```diff
@@ -310,7 +310,7 @@ impl xtra::Actor for Actor {
         self.tasks.add(
             this.clone()
-                .send_interval(std::time::Duration::from_secs(5), || Sync),
+                .send_interval(std::time::Duration::from_secs(30), || Sync),
```
🤔 Should we have two intervals?

We know that we only get one attestation per hour, but it is very important to get attestations as soon as possible. We could optimise this further but I think it's acceptable to check for a new attestation every 30 seconds.

But the announcements are updated at a completely different rate. We look 25 hours in the future, but we could look months ahead (e.g. https://outcome.observer/h00.ooo/x/BitMEX/BXBT/2022-07-20T01:00:00.price?n=20).

I would recommend to look 48 hours ahead and only run `ensure_having_announcements` once per hour. Would leave a lot of space for further optimisation, but the gains would be small given that we would already be doing

https://github.com/itchysats/itchysats/blob/5e6d8e63c1636eb6d7de725b6c489622e3da466b/daemon/src/oracle.rs#L136-L141

48 times per hour, which is pretty inexpensive.

bonomat replied:
I think it is acceptable to check every 30 seconds for now. We can reduce this to check even less often but I'm not sure how much would gain from this.

luckysori replied:
Sure, we don't have to go to 60 minutes now.

But why 30 seconds for announcements and 60 seconds for attestations? It would make more sense the other way around for me.

bonomat replied:
Fair point. what timing would you chose?

luckysori replied:
Maybe:

- Attestations every 30 seconds.
- Announcements every minute.

bonomat replied:
works for me :)

luckysori commented on daemon/src/lib.rs:
```diff
@@ -90,7 +90,7 @@ impl<O, W, P> TakerActorSystem<O, W, P>
     O: Handler<oracle::MonitorAttestation>
         + Handler<oracle::GetAnnouncement>
-        + Handler<oracle::Sync>
+        + Handler<oracle::SyncAnnouncements>
```
🔧 You can remove this altogether (in a separate commit? :). `oracle::Sync` should be a module-private type, but we made it public to use it in `daemon-tests`. The `TakerActorSystem` doesn't need to know about the internal messages `oracle::Actor` sends to itself (verified locally).

luckysori commented on maker/src/actor_system.rs:
```diff
@@ -55,7 +55,7 @@ impl<O, W> ActorSystem<O, W>
     O: Handler<oracle::MonitorAttestation>
         + Handler<oracle::GetAnnouncement>
-        + Handler<oracle::Sync>
+        + Handler<oracle::SyncAnnouncements>
```
Same thing I mentioned for `TakerActorSystem` applies here (also tested).

luckysori reviewed (approved):
I left a comment before seeing your last commit 😛 I like the changes, although I would tweak the timing of the internal messages.

bonomat committed:
TakerActorSystem does not need to know about oracle sync types

bonomat committed:
Increase sync time

There is no need to sync every 5 seconds as we expect to fetch a new announcement only 1/h.
Syncing more often might unnecessarily use CPU cycles or even fill up the oracle actor's inbox with messages.

bonomat committed:
Split sync attestations and announcements

We only care about fetching new attestations after the initial load. Because, when a CFD is loaded from the DB, we add the needed `Attestation` (using `MonitorAttestation` msg) to a list which will then be fetched from the oracle on every `SyncAttestations` interval. Given initial loading from the DB takes some time, we start syncing only after the loading is complete.
